ComponentKit/DataSources/Common/CKComponentDataSourceAttachController.mm

```diff
@@ -0,0 +1,170 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import <objc/runtime.h>
+#import "CKComponentInternal.h"
+
+#import "CKComponentDataSourceAttachController.h"
+
+@interface CKComponentDataSourceAttachState : NSObject
+
+@property (nonatomic, strong, readonly) NSSet *mountedComponents;
+@property (nonatomic, readonly) CKComponentScopeRootIdentifier scopeIdentifier;
+
+- (instancetype)initWithScopeIdentifier:(CKComponentScopeRootIdentifier)scopeIdentifier
+                      mountedComponents:(NSSet *)mountedComponents;
+@end
+
+@interface UIView(CKComponentDataSourceAttachController)
+
+@property (nonatomic, strong, setter=ck_setAttachState:) CKComponentDataSourceAttachState *ck_attachState;
+
+@end
+
+@implementation CKComponentDataSourceAttachController
+{
+  /**
+   We keep a strong reference to the mounted view to enforce that every view
+   being attached will have to be detached before the view is deallocated.
+   */
+  NSMutableDictionary *_scopeIdentifierToAttachedViewMap;
+}
+
+#pragma mark - Initialization/Teardown
+
+- (instancetype)init
+{
+  self = [super init];
+  if (self) {
+    _scopeIdentifierToAttachedViewMap = [NSMutableDictionary dictionary];
+  }
+  return self;
+}
+
+- (void)dealloc
+{
+  NSEnumerator *viewsEnumerator = _scopeIdentifierToAttachedViewMap.objectEnumerator;
+  void(^deallocBlock)(void) = ^{
+    for (UIView *attachedView in viewsEnumerator) {
+      [self _detachComponentLayoutFromView:attachedView];
```

Comment:
We probably shouldn't capture self in a scope that will live longer than the life of dealloc.  Can we use a static C function instead?